[BD-13] Deprecates remaining user-related properties on ModuleSystem

pomegranited opened the pull request:
## Description

Builds on the shim added by https://github.com/edx/edx-platform/pull/29190 to deprecate these remaining user-related `ModuleSystem` attributes in favour of the user service:

* user_location
* get_real_user
* get_user_role

These properties are only used by a few XBlocks: LTI, LTI2.0, Video (CDN/branding), edxnotes

This change is only a refactoring, and should not affect Learners, Course Authors, or anyone else using edx-platform.

## Supporting information

* [Content Core Platform Simplication project](https://openedx.atlassian.net/wiki/spaces/AC/pages/3091629503/Content+Core+Platform+Simplification)
* [BLENDED-113](https://openedx.atlassian.net/browse/BLENDED-113) (edX internal ticket)
* [FAL-2399](https://tasks.opencraft.com/browse/FAL-2399) (OpenCraft internal ticket)

## Testing instructions

* LMS: https://pr29375.sandbox.opencraft.hosting/
* Studio: https://studio.pr29375.sandbox.opencraft.hosting/

Basic regression testing can be run on the sandbox.

To test the LTI 2.0 (consumer) component: 
1. View the [LTI component in Studio](https://studio.pr29375.sandbox.opencraft.hosting/container/block-v1:OpenCraft+BD-13+2021_1+type@vertical+block@0a1f09b607794a42adefde02b49184f4).
   Note that the LTI URL points to a problem block on a different sandbox server's Demo course.
 1.  Ensure you can edit the component, preview it, view it in the LMS, and submit answers to the problems shown.

To test edx-notes:

1. Login to LMS as a student user.
1. View the BD-13 test course in the LMS.
2. Navigate to an [HTML block](https://discovery.pr29375.sandbox.opencraft.hosting/learning/course/course-v1:OpenCraft+BD-13+2021_1/block-v1:OpenCraft+BD-13+2021_1+type@sequential+block@c2883d06799c4aaf845d6abd4f934328/block-v1:OpenCraft+BD-13+2021_1+type@vertical+block@a516543787854cb2903ffd1c2a642f8e), select some text, and ensure you can add/edit notes.
3. Toggle the hide/show notes button in the lower right corner to hide/show notes.
2. Click on the [Notes](https://pr29375.sandbox.opencraft.hosting/courses/course-v1:OpenCraft+BD-13+2021_1/edxnotes/) tab and ensure your notes are visible there.

To test the video CDN/branding:

You need to view the site from one of the CDN-configured countries, currently Australia (`AU`), Pakistan (`PK`), or Poland (`PL`).

Note: I don't really have a valid CDN to use here, but enabling a CDN allows a [configured branding logo to be shown under the video](https://github.com/edx/edx-platform/blob/8b3d7e040c2d3b138e443466c83ed3dc7adcb94a/common/lib/xmodule/xmodule/video_module/video_module.py#L336-L337), so this at least is visible and can be tested.

1. View the [test Video in the LMS](https://pr29375.sandbox.opencraft.hosting/courses/course-v1:OpenCraft+BD-13+2021_1/jump_to/block-v1:OpenCraft+BD-13+2021_1+type@vertical+block@3f7e6c9b074c4bb7b4a2f6c7aa6b1d2c).
   Note that the configured branding logo is shown under the video when you view this content from one of the configured countries.
 
## Deadline

None

## Author Notes & Concerns

Setup for user location/CDN testing:

1. Update `EDXAPP_VIDEO_CDN_URLS` or `/edx/etc/lms.yml` to provide a CDN URL for a given country code (doesn't matter what the URL is), eg.
   ```yaml
   VIDEO_CDN_URL:
     AU: http://example.com/edx/video?s3_url=
   ```
1. As a superuser, navigate to Django Admin > Branding > [Branding Info Config](https://pr29375.sandbox.opencraft.hosting/admin/branding/brandinginfoconfig/), and add a branding url, logo, and logo_tag for your country, e.g.
   ```json
   {
      "AU": {
        "url": "https://www.sa.gov.au",
        "logo_src": "https://www.sa.gov.au/__data/assets/image/0012/251013/sa-logo.png",
        "logo_tag": "Logo for South Australia"
      }
    }
    ```

edx-webhook commented:
Thanks for the pull request, @pomegranited! I've created [BLENDED-1020](https://openedx.atlassian.net/browse/BLENDED-1020) to keep track of it in Jira.   More details are on the [BD-13](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3062333478/BD-13+xModule+--+xBlock+Conversion+Phase+2) project page.


When this pull request is ready, tag your edX technical lead.


<!-- comment:welcome-blended -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

symbolist commented on common/lib/xmodule/xmodule/video_module/video_module.py:
```diff
@@ -280,7 +282,8 @@ def get_html(self, view=STUDENT_VIEW):  # lint-amnesty, pylint: disable=argument
         # we leverage a geography specific CDN, like China.
         default_cdn_url = getattr(settings, 'VIDEO_CDN_URL', {}).get('default')
-        cdn_url = getattr(settings, 'VIDEO_CDN_URL', {}).get(self.system.user_location, default_cdn_url)
+        user_location = self.system.service(self, 'user').get_current_user().opt_attrs[ATTR_KEY_REQUEST_COUNTRY_CODE]
```
Could this be `None`?

pomegranited replied:
Documentation says it could be an empty string, but not `None`.  This string is stored in the session in the geioinfo middleware, pulled from [openedx.core.djangoapps.geoinfo.api.country_code_from_ip](https://github.com/edx/edx-platform/blob/b1fa2bdae628d7b8195e4069282f3ee76e12136f/openedx/core/djangoapps/geoinfo/api.py#L18-L19).

Why? I don't see anything in this code that makes `None` a risk?

symbolist replied:
Got it. Don't really remember what I was thinking. :)

symbolist commented on common/lib/xmodule/xmodule/tests/test_lti_unit.py:
```diff
@@ -173,7 +172,7 @@ def test_real_user_is_none(self):
         self.xmodule.verify_oauth_body_sign = Mock()
         self.xmodule.has_score = True
-        self.system.get_real_user = Mock(return_value=None)
+        self.system.service(self.xmodule, 'user').user = None
```
Hmm, does the `UserService` have a `user` property?

pomegranited replied:
No -- this was just my shortcut to getting a `None` user returned by the `StubUserService.get_user_by_anonymous_id`. I'll find a better way.

pomegranited commented on common/djangoapps/xblock_django/user_service.py:
```diff
@@ -35,11 +36,16 @@ def __init__(self, django_user, **kwargs):
         self._anonymous_user_id = kwargs.get('anonymous_user_id', None)
+        self._request_country_code = kwargs.get('request_country_code', None)
+
+        self.get_user_role = kwargs.get('get_user_role', lambda: 'student')
```
Nope.. I thought maybe it was better to call the method only when it was needed, but since all that access stuff is cached, there's no risk to do that here. Fixed with 831393e .

symbolist commented on common/lib/xmodule/xmodule/x_module.py:
```diff
@@ -1809,6 +1811,56 @@ def user_is_staff(self):
+        return None
+
+    @property
+    def get_user_role(self):
```
Is this supposed to return a function?

pomegranited replied:
Oops yes, yes it is, I messed that up with 831393e .. will fix that.

Agrendalath commented on common/lib/xmodule/xmodule/tests/helpers.py:
```diff
@@ -74,10 +81,19 @@ def get_current_user(self):
         user = XBlockUser()
         if self.user.is_authenticated:
```
We're not mocking the user anymore, so this is going to throw an `AttributeError` if the user is `None`.

I see that we don't want to mock the user because of the test in `common/lib/xmodule/xmodule/tests/test_lti_unit.py`. Therefore, let's add a safety check (`is not None`), to avoid throwing exceptions for default values of the class attributes.

Agrendalath commented on common/lib/xmodule/xmodule/tests/test_lti20_unit.py:
```diff
@@ -20,9 +20,8 @@ class LTI20RESTResultServiceTest(unittest.TestCase):
     def setUp(self):
         super().setUp()
-        self.system = get_test_system()
+        self.system = get_test_system(user=self.USER_STANDIN)
```
Nit: let's define [USER_STANDIN](https://github.com/edx/edx-platform/blob/185bedc707e368c5eb4a874acdaccd8b341d30be/common/lib/xmodule/xmodule/tests/test_lti20_unit.py#L225-L226) before its first usage. It's easier to read the code this way.

Agrendalath commented on common/lib/xmodule/xmodule/tests/test_lti_unit.py:
```diff
@@ -171,9 +171,9 @@ def test_real_user_is_none(self):
         """
         If we have no real user, we should send back failure response.
         """
+        self.system._services['user'] = StubUserService(user=None)  # pylint: disable=protected-access
```
Why can't we use `self.system.service(self.xmodule, 'user').user = None` here, like in the test from `common/lib/xmodule/xmodule/tests/test_lti20_unit.py`?

pomegranited replied:
Actually, they should both be using the `StubUserService` mechanism instead, because the real `DjangoXBlockUserService` class doesn't have a `user` attribute. see https://github.com/edx/edx-platform/pull/29375#discussion_r757109450 for discussion.

Good catch, I missed this one!

pomegranited commented:
Thanks for your detailed review @Agrendalath ! I've addressed your points, and [deployed a new appserver](https://manage.opencraft.com/instance/31800/). Could you activate that one and use it for your testing?

Also, I'd like to squash my commits before merging, so let me know when you're ready for me to do that?

Agrendalath reviewed (approved):
:+1:

* [x]  I tested this: verified that the refactored properties are working correctly
* [x]  I read through the code
* [x]  I checked for accessibility issues: n/a
* [x]  Includes documentation
* [x] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository: n/a

Agrendalath commented:
@pomegranited, thank you for preparing the sandbox instance and nice testing instructions! Feel free to squash the commits.

I've noticed an odd behavior of the LTI provider, though it doesn't seem to be related to this PR. When I"m opening the [LTI XBlock in the LMS](https://discovery.pr29375.sandbox.opencraft.hosting/learning/course/course-v1:OpenCraft+BD-13+2021_1/block-v1:OpenCraft+BD-13+2021_1+type@sequential+block@862667c509144a4c83381589185f9072/block-v1:OpenCraft+BD-13+2021_1+type@vertical+block@0a1f09b607794a42adefde02b49184f4) as a "staff" user  - it's working correctly. However, when I'm using the "audit", then I'm getting an error page. There is the following traceback on the `pr29260` instance:
```python
Traceback (most recent call last):
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/lti_provider/users.py", line 31, in authenticate_lti_user
    lti_user = LtiUser.objects.get(
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/db/models/query.py", line 435, in get
    raise self.model.DoesNotExist(
lms.djangoapps.lti_provider.models.LtiUser.DoesNotExist: LtiUser matching query does not exist.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/newrelic/hooks/framework_django.py", line 562, in wrapper
    return wrapped(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/edx/app/edxapp/edx-platform/common/djangoapps/util/views.py", line 188, in inner
    response = view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/lti_provider/views.py", line 91, in lti_launch
    authenticate_lti_user(request, params['user_id'], lti_consumer)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/lti_provider/users.py", line 37, in authenticate_lti_user
    lti_user = create_lti_user(lti_user_id, lti_consumer)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/lti_provider/users.py", line 59, in create_lti_user
    edx_user = User.objects.create_user(
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/contrib/auth/models.py", line 152, in create_user
    return self._create_user(username, email, password, **extra_fields)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/contrib/auth/models.py", line 146, in _create_user
    user.save(using=self._db)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/contrib/auth/base_user.py", line 67, in save
    super().save(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/db/models/base.py", line 726, in save
    self.save_base(using=using, force_insert=force_insert,
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/db/models/base.py", line 774, in save_base
    post_save.send(
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/dispatch/dispatcher.py", line 180, in send
    return [
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/dispatch/dispatcher.py", line 181, in <listcomp>
    (receiver, receiver(signal=self, sender=sender, **named))
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/enterprise/decorators.py", line 88, in wrapper
    signal_handler(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/enterprise/signals.py", line 68, in handle_user_post_save
    for pending_ecu in pending_ecus:
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/db/models/query.py", line 280, in __iter__
    self._fetch_all()
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/db/models/query.py", line 1324, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/db/models/query.py", line 51, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1175, in execute_sql
    cursor.execute(sql, params)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/db/backends/utils.py", line 78, in _execute
    self.db.validate_no_broken_transaction()
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/db/backends/base/base.py", line 447, in validate_no_broken_transaction
    raise TransactionManagementError(
django.db.transaction.TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.
```

pomegranited commented:
Thanks for your review @Agrendalath ! I've rebased and squashed my commits, so this should be ready to merge once tests pass again.

> I've noticed an odd behavior of the LTI provider, though it doesn't seem to be related to this PR. 

I agree, that's odd, but doesn't look related to this PR at all, since it's happening on the other server for https://github.com/edx/edx-platform/pull/29260. It seems to have something to do with some post-save signals attached to the User model, but I can't tell for sure. It wasn't happening with a previous version of edx-platform on that server -- I'll investigate it some more on that PR.

pomegranited commented:
@ormsbee Are you ok with @Agrendalath merging this PR (and https://github.com/edx/edx-platform/pull/29312), or would you rather deploy them on your timeline? I don't want to mess up anyone's holiday :) But it would be great to get these in soon.

Agrendalath commented:
@jristau1984, I'm merging this, as discussed on Slack.

edx-webhook commented:
@pomegranited 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/29375&entry.1484655318=https://github.com/pomegranited&entry.752974735=2021-11-22+07:49&entry.1917517419=2021-12-20+14:41&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
